linux/ibus-keyman/src/engine.c

```diff
@@ -613,55 +661,116 @@ ibus_keyman_engine_process_key_event (IBusEngine     *engine,
+                    g_message("removing one utf8 char from CHAR buffer");
+                    glong end_pos = g_utf8_strlen(keyman->char_buffer, -1);
+                    gchar *new_buffer;
+                    if (end_pos == 1) {
+                        new_buffer = NULL;
+                        g_message("resetting CHAR buffer to NULL");
+                    }
+                    else {
+                        new_buffer = g_utf8_substring(keyman->char_buffer, 0 , end_pos - 1);
+                        g_message("changing CHAR buffer to :%s:", new_buffer);
+                    }
+                    if (g_strcmp0(keyman->char_buffer, new_buffer) == 0) {
+                        g_message("oops, CHAR buffer hasn't changed");
+                    }
+                    g_free(keyman->char_buffer);
+                    keyman->char_buffer = new_buffer;
+                }
+                else if (ok_for_single_backspace(action_items, i, num_action_items)) {
+                    // single backspace can be handled by ibus as normal
+                    g_message("no char actions, just single back");
+                    return FALSE;
+                }
+                else {
+                    g_message("DAR: ibus_keyman_engine_process_key_event - client_capabilities=%x, %x", engine->client_capabilities,  IBUS_CAP_SURROUNDING_TEXT);
+
+                    if ((engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT) != 0) {
+                        g_message("deleting surrounding text 1 char");
+                        ibus_engine_delete_surrounding_text(engine, -1, 1);
+                    } else {
+                        g_message("forwarding backspace with reset context");
+                        km_kbp_context_item *context_items;
+                        km_kbp_context_get(km_kbp_state_context(keyman->state),
+                            &context_items);
+                        reset_context(engine);
+                        forward_keycode(keyman, KEYMAN_BACKSPACE, 0);
+                        km_kbp_context_set(km_kbp_state_context(keyman->state),
+                            context_items);
+                        km_kbp_context_items_dispose(context_items);
+                    }
                 }
                 break;
             case KM_KBP_IT_PERSIST_OPT:
-                g_message("PERSIST_OPT action");
+                g_message("PERSIST_OPT action %d/%d", i+1, (int)num_action_items);
                 break;
             case KM_KBP_IT_EMIT_KEYSTROKE:
-                g_message("EMIT_KEYSTROKE action");
+                if (keyman->char_buffer != NULL)
+                {
+                    ibus_keyman_engine_commit_string(keyman, keyman->char_buffer);
+                    g_free(keyman->char_buffer);
+                    keyman->char_buffer = NULL;
+                }
+                g_message("EMIT_KEYSTROKE action %d/%d", i+1, (int)num_action_items);
                 return FALSE;
```

Comment:
This makes the assumption that `KM_KBP_IT_EMIT_KEYSTROKE` will be the final action. That's not explicitly documented anywhere so it may be better to make it a cached return value. e.g. `KM_KBP_IT_INVALIDATE_CONTEXT` could in theory occur after this, or even other actions, even if it is more logical to have the emit keystroke at the end.